### DISCUSSION TITLE
WIT support via `wit-bindgen`

### DISCUSSION THREAD

**Discussion by flavio** [wasmi-labs/wasmi#703] [Show and tell]
> Note: this was initially filed as an [issue](https://github.com/paritytech/wasmi/issues/657)

As part of the [WebAssembly Component Model](https://github.com/WebAssembly/component-model) proposal, the [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md) format is used to describe the interfaces used by components.

There's a cli tool, [`wit-bindgen`](https://github.com/bytecodealliance/wit-bindgen) that can to scaffold the import/export code of a certain component.

Currently wit-bindgen can generate export and import code for Wasmtime, but it lacks wasmi support.

In case you're interested, I've created a forked version of wit-bindgen that can generate both import and export code for wasmi.
The code currently lives inside of [this repository](https://github.com/flavio/wit-bindgen/tree/wasmi), under the `wasmi` branch.

wit-bindgen is undergoing a lot of changes. I've targeted the `v0.2.0` release, which is the one used by projects like [Spin](https://github.com/fermyon/spin) and [SpiderLightning](https://github.com/deislabs/spiderlightning).

I will not push the code upstream yet, because of all these undergoing changes (see [this comment](https://github.com/bytecodealliance/wit-bindgen/issues/306#issuecomment-1335854999)).

I'm opening this issue in case other wasmi users could benefit from using this code.

If you want to see the code in action, you can take a look at [this](https://github.com/flavio/hermit-wasm) codebase.

**Comment by Robbepop** [comment]
Let me repeat again how cool this is: 🚀🎉 
Looking forward to the update that uses the new `Linker` API for linker based host functions. :)

**Reply by flavio** [reply]
Good news, I've just updated the `wit-bindgen` fork to make use of the `Linker::func_wrap` function introduced by the `0.28.0` release :partying_face:

**Reply by Robbepop** [reply]
Hey thanks for letting us know! Have there been any problems or annoyances while using the new API?

**Reply by flavio** [reply]
Nope, everything worked smoothly 😄

**Comment by jakehemmerle** [comment]
Nice work!

We're building on substrate, and we're also doing some other Wasm stuff off-chain. So far we've ruled out using wasmi for the offchain part because of no WIT support, but will be reconsidering it if this gets merged, since interpreter is ideal, and could give us the flexability to do WIT on-chain later.

Thanks for the efforts here, I'll be keeping eyes on this! :)

**Comment by Robbepop** [comment]
I just pinned this discussion for people to get note of it.

See https://github.com/paritytech/wasmi/discussions/723 for a usage example with JS types.
